### DISCUSSION TITLE
Whats the best way to cache a CLI?

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [actions/cache#386] [General]
Part of my action uses the IBM Cloud CLI (`ibmcloud`) and it takes forever to install, so I'd like to cache it.  I tried to use `which` to figure out where the executables were going, but I've run into a few snags...

Here's what I have so far within my job (should refresh the cache at the start of each week):
```
- name: Get CLI Cache Key
  id: get-cache-key
  run: |
    echo "::set-output name=cache-key::$(/bin/date -u "+%Y%U")"
  shell: bash

- uses: actions/cache@v2
  id: cache
  with:
    path: |
      /home/runner/.bluemix
      /usr/local/bin/ibmcloud
      /usr/bin/kubectl
    key: ${{ runner.os }}-${{ steps.get-cache-key.outputs.cache-key }} # CLI cache becomes stale every week
```

When I run this Action, however, I get the following error: 
```
Run actions/cache@v2
Received 20034758 of 20034758 (100.0%), 52.5 MBs/sec
Cache Size: ~19 MB (20034758 B)
/bin/tar --use-compress-program zstd -d -xf /home/runner/work/_temp/****-****-****/cache.tzst -P -C /home/runner/work/****/****
/bin/tar: ../../../../../usr/local/bin/ibmcloud: Cannot create symlink to ‘/usr/local/ibmcloud/bin/ibmcloud’: Permission denied
/bin/tar: ../../../../../usr/bin/kubectl: Cannot open: File exists
/bin/tar: Exiting with failure status due to previous errors
[warning]Tar failed with error: The process '/bin/tar' failed with exit code 2
```

Any ideas on what I'm doing wrong or how this could be improved?

**Comment by dhadka** [comment]
Something to try is changing the installation path to somewhere in the working directory.  That should help avoid permission issues.  See https://cloud.ibm.com/docs/cli?topic=cli-install-ibmcloud-cli#install-custom-dir

**Reply by SpencerKaiser** [reply]
That's a great idea, but I'm not sure how I'd go about that programmatically from within an Action. It sounds like the only way to do that is by manually downloading the binary and moving it where you'd like. It also doesn't look like their installer script provides a way to install to a custom directory 😓

**Reply by dhadka** [reply]
Here's a workflow I put together that sets up the cli: https://github.com/dhadka/ibmcloud-example/blob/master/.github/workflows/blank.yml

And the corresponding run: https://github.com/dhadka/ibmcloud-example/runs/924238414?check_suite_focus=true

Given this takes ~4 seconds to setup, I don't think caching is necessary.

**Reply by dhadka** [reply]
...but since this is the `actions/cache` repo, if you did want to cache the tool and save 2 seconds, here's an example: https://github.com/dhadka/ibmcloud-example/blob/master/.github/workflows/with_cache.yml 😄 

And the run: https://github.com/dhadka/ibmcloud-example/actions/runs/187460009

**Reply by SpencerKaiser** [reply]
The CLI itself is tiny, but I lose like 30 seconds on the sub-dependencies (mainly `kubectl`)... which I'm not sure how to cache those either 😓 that example is super helpful though, so maybe I can follow that pattern for the others.

**Reply by dhadka** [reply]
Yeah, seems like `kubectl` can be setup similarly - https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl-on-linux

**Reply by SpencerKaiser** [reply]
I just realized the one downside to this pattern is that it's version locked 😔 I had a cache key setup where the cache would expire every week but with this pattern I'd need to make a change to the workflow to update them. On the upside the cache would last longer...

**Reply by SpencerKaiser** [reply]
So I guess that's one tiny feature request I'd make for Actions: I would love to be able to cache system directories within the runners. Ideally when installing a package or something, I'd be able to cache it wherever it's located instead of only at the working directory

**Reply by SpencerKaiser** [reply]
@dhadka I ended up bailing on the idea; it seems too difficult to cache the CLI with plugins (I can't even figure out how to install the plugins to a custom directory anyways), but I'm going to mark your answer as correct since you solved the original problem of caching the CLI on it's own.

For the future, I would love to see GitHub provide the ability to cache non-working directory files to make things easier for the end user 🙂

**Reply by kaczmarj** [reply]
@dhadka - thank you very much, your answer clears things up. for others reading and looking at the code examples, `add-path` has been deprecated in favor of writing paths to `$GITHUB_PATH`. See [the documentation](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#adding-a-system-path).

```diff
- echo "::add-path::$(pwd)/IBM_Cloud_CLI"
+ echo "$(pwd)/IBM_Cloud_CLI" >> $GITHUB_PATH
```

**Comment by samus-aran** [comment]
We're locking this discussion because it has not had recent activity and/or other members have asked for more information to assist you but received no response. Thank you for helping us maintain a productive and tidy community for all our members.
